cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -2871,6 +2872,12 @@ Status S3FileSystem::CreateDir(const std::string& s, bool recursive) {
     for (const auto& part : path.key_parts) {
       parent_key += part;
       parent_key += kSep;
+      if (options().check_directory_existence_before_creation) {
```

Comment:
I just reflect on it and my 2cents is to keep walking down logic.  I'm doing the checking inside the creation step. I can only build a wall from the bottom up otherwise the world is upside down :)  So I must check `a` then `a/b` and `a/b/c`.

Reply:
What I'm proposing is (example in the case of `a/b/c`):
1. check which ancestors exist by walking _up_ the directory tree: first `a/b/c` then `a/b`... until you find the first existing ancestor
2. create the missing descendents by walking _down_ the directory tree: for example, if you just found that `a` exists, create `a/b` then `a/b/c`

The idea is that, most of the time, almost the entire directory chain will exist, especially if your workload is writing into a deeply partitioned dataset. So doing the directory checks from leaf to root should issue less requests and have less latency.

More formally, let's call _n_ the depth of the path given to `CreateDir`, and _m_ the number of directories missing along that path.
* with the current approach, we're calling `HeadObject` _O(n)_ times and `PutObject` _O(m)_ times;
* with my proposal, we're calling `HeadObject` _O(m)_ times and `PutObject` _O(m)_ times.

Assuming that _m_ is on average much smaller than _n_ (_m_ would be 0 or 1 most of the time), the approach I'm proposing should be much more efficient, and it would never be less efficient anyway.

Does that make sense or am I missing something?

Reply:
@pitrou Could you please review again?
In this round:
* I added the above logic.
* Ensure my tests are working as I foget to call `MakeFileSystem()`
* Add missing tests for non recursive CreateDir scenario.